core/src/main/scala/kafka/server/RaftRequestHandler.scala

```diff
@@ -48,6 +56,58 @@ class RaftRequestHandler(networkChannel: KafkaNetworkChannel,
             requestBody,
             response => sendResponse(request, Some(response)))
 
+        case ApiKeys.API_VERSIONS =>
+          sendResponse(request, Option(ApiVersionsResponse.apiVersionsResponse(0, 2)))
+
+        case ApiKeys.METADATA =>
+          val topics = new MetadataResponseData.MetadataResponseTopicCollection
+          val leaderConnection = networkChannel.getConnectionInfo(counter.leaderId())
+          if (leaderConnection != null) {
+            topics.add(new MetadataResponseTopic()
+              .setErrorCode(Errors.NONE.code)
+              .setName(metadataPartition.topic())
+              .setIsInternal(true)
+              .setPartitions(Collections.singletonList(new MetadataResponsePartition()
+                .setErrorCode(Errors.NONE.code())
+                .setPartitionIndex(metadataPartition.partition())
+                .setLeaderId(leaderConnection.id()))))
+          }
+
+          val brokers = new MetadataResponseData.MetadataResponseBrokerCollection
+
+          networkChannel.allConnections().foreach( connection => {
+            brokers.add(new MetadataResponseData.MetadataResponseBroker()
+              .setNodeId(connection.id())
+              .setHost(connection.host())
+              .setPort(connection.port()))
+          })
+
+          sendResponse(request, Option(new MetadataResponse(
+            new MetadataResponseData()
+              .setTopics(topics)
+              .setBrokers(brokers))))
+
+        case ApiKeys.PRODUCE =>
+          val produceRequest = request.body[ProduceRequest]
+          if (!counter.isInstanceOf[NoOpReplicatedCounter]) {
+            throw new IllegalStateException("Should use a no-op replicated counter for produce")
+          }
+
+          counter.asInstanceOf[NoOpReplicatedCounter].appendRecords(
```

Comment:
We would return the future from appender once the local append completes, which means it is effectively always "ack == 1". Is that intentional? I thought we are comparing Raft with "ack == quorum" with Kafka where "ack == all".

Reply:
Good question, I didn't check this case. Do we have a ticket to introduce `ack` for state machine apply? I could address this comment in a separate PR if you see fit.

Reply:
Filed: https://confluentinc.atlassian.net/browse/KC-660